pkg/util/voljournal.go

```diff
@@ -0,0 +1,383 @@
+that we do not use a uuid that is already in use
+
+- Next, a key with the VolName is created in the csiDirectory, and its value is updated to store the
+generated uuid
+
+- This is followed by updating the CephUUIDDirectory with the VolName in the csiNameKey
+
+- Finally, the volume is created (or promoted from a snapshot, if content source was provided),
+using the uuid and a corresponding name prefix (namingPrefix) as the volume name
+
+The entire operation is locked based on VolName hash, to ensure there is only ever a single entity
+modifying the related omaps for a given VolName.
+
+This ensures idempotent nature of creates, as the same CO generated VolName would attempt to use
+the same volume uuid to serve the request, as the relations are saved in the respective omaps.
+
+Deletion of omaps:
+Delete requests would not contain the VolName, hence deletion uses the volume ID, which is encoded
+with the volume uuid in it, to find the volume and the CephUUIDDirectory. The CephUUIDDirectory is
+read to get the VolName that this image points to. This VolName can be further used to read and
+delete the key from the csiDirectory.
+
+As we trace back and find the VolName, we also take a hash based lock on the VolName before
+proceeding with deleting the volume and the related omap entries, to ensure there is only ever a
+single entity modifying the related omaps for a given VolName.
+*/
+
+type CSIJournal struct {
+	// csiDirectory is the name of the CSI volumes object map that contains CSI volume-name (or
+	// snapshot name) based keys
+	csiDirectory string
+
+	// CSI volume-name keyname prefix, for key in csiDirectory, suffix is the CSI passed volume name
+	csiNameKeyPrefix string
+
+	// Per Ceph volume (RBD/FS-subvolume) object map name prefix, suffix is the generated volume uuid
+	cephUUIDDirectoryPrefix string
+
+	// CSI volume-name key in per Ceph volume object map, containing CSI volume-name for which the
+	// Ceph volume was created
+	csiNameKey string
+
+	// source volume name key in per Ceph snapshot object map, containing Ceph source volume uuid
+	// for which the snapshot was created
+	cephSnapSourceKey string
+
+	// volume name prefix for naming on Ceph rbd or FS, suffix is a uuid generated per volume
+	namingPrefix string
+}
+
+var (
+	CSIVolumeKeyNames = CSIJournal{
+		csiDirectory:            "csi.volumes",
+		csiNameKeyPrefix:        "csi.volume.",
+		cephUUIDDirectoryPrefix: "csi.volume.",
+		csiNameKey:              "csi.volname",
+		namingPrefix:            "csi-vol-",
+		cephSnapSourceKey:       "",
+	}
+
+	CSISnapshotKeyNames = CSIJournal{
+		csiDirectory:            "csi.snaps",
+		csiNameKeyPrefix:        "csi.snap.",
+		cephUUIDDirectoryPrefix: "csi.snap.",
+		csiNameKey:              "csi.snapname",
+		namingPrefix:            "csi-snap-",
+		cephSnapSourceKey:       "csi.source",
+	}
+)
+
+// CSIVolumeJournal returns an instance of volume keys
+func CSIVolumeJournal() CSIJournal {
```

Comment:
instead of declaring `CSISnapshotKeyNames` and `CSIVolumeKeyNames` outside  of function
we can create a new function as `NewCSIVolumeJournal` which fills the struct and return it same applies for snapshot also

Reply:
Done